app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt

```diff
@@ -0,0 +1,224 @@
+package org.stepic.droid.code.ui
+
+import android.content.Context
+import android.graphics.*
+import android.support.v7.widget.AppCompatEditText
+import android.text.*
+import android.text.style.ForegroundColorSpan
+import android.util.AttributeSet
+import android.util.Log
+import android.view.View
+import android.widget.ScrollView
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.disposables.Disposable
+import io.reactivex.schedulers.Schedulers
+import io.reactivex.subjects.BehaviorSubject
+import io.reactivex.subjects.PublishSubject
+import org.stepic.droid.code.highlight.prettify.PrettifyParser
+import org.stepic.droid.code.highlight.syntaxhighlight.ParseResult
+import org.stepic.droid.code.highlight.themes.CodeTheme
+import org.stepic.droid.code.highlight.themes.Presets.DefaultTheme
+import org.stepic.droid.util.DpPixelsHelper
+import java.util.concurrent.TimeUnit
+
+class CodeEditor : AppCompatEditText, TextWatcher {
+    companion object {
+        const val SCROLL_DEBOUNCE_MS = 100L
+        const val INPUT_DEBOUNCE_MS = 300L
+    }
+
+    constructor(context: Context) : super(context)
+    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
+    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)
+
+    private val parser = PrettifyParser()
+
+    private val LINE_NUMBERS_MARGIN by lazy {
+        DpPixelsHelper.convertDpToPixel(4f).toInt()
+    }
+
+    private val highlightPublisher = PublishSubject.create<Editable>()
+    private val spanPublisher = BehaviorSubject.create<List<ParseResult>>()
+    private val scrollPublisher = PublishSubject.create<Int>()
+    private var scrollDisposable: Disposable? = null
+
+    private val compositeDisposable = CompositeDisposable()
+
+    private val lineNumbersBackgroundPaint = Paint()
+    private val lineNumbersStrokePaint = Paint()
+    private val selectedLinePaint = Paint()
+    private val lineNumbersTextPaint by lazy {
+        val p = Paint()
+        p.typeface = Typeface.MONOSPACE
+        p.textAlign = Paint.Align.RIGHT
+        p.textSize = textSize * 0.8f
+        p.flags = p.flags or Paint.ANTI_ALIAS_FLAG
+        p
+    }
+
+    var lang = ""
+        set(value) {
+            field = value
+            afterTextChanged(editableText) // refresh highlight
+        }
+
+    
+    var scrollContainer: ScrollView? = null
```

Comment:
what if we will set this value more than 1 time? ScrollPublisher will publish events to old observer (memory leak)

Reply:
We need to dispose old observer (if it is existed) and after that subscribe new

Reply:
Why we should set `scrollContainer`, if edit text is scrollable? Check `Scroller` class